dot/core/consensus.go

```diff
@@ -0,0 +1,241 @@
+	authData := s.bs.AuthorityData()
+	if len(authData) == 0 {
+		return nil, fmt.Errorf("authority data not set")
+	}
+
+	newBlocks := make(chan types.Block)
+	s.blkRec = newBlocks
+
+	epochDone := make(chan struct{})
+	s.epochDone = epochDone
+
+	babeKill := make(chan struct{})
+	s.babeKill = babeKill
+
+	keys := s.keys.Sr25519Keypairs()
+
+	// get best slot to determine next start slot
+	bestHash := s.blockState.BestBlockHash()
+	bestSlot, err := s.blockState.GetSlotForBlock(bestHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get slot for latest block: %s", err)
+	}
+
+	// BABE session configuration
+	bsConfig := &babe.SessionConfig{
+		Keypair:          keys[0].(*sr25519.Keypair),
+		Runtime:          s.rt,
+		NewBlocks:        newBlocks, // becomes block send channel in BABE session
+		BlockState:       s.blockState,
+		StorageState:     s.storageState,
+		TransactionQueue: s.transactionQueue,
+		AuthData:         authData,
+		Done:             epochDone,
+		Kill:             babeKill,
+		StartSlot:        bestSlot + 1,
+		SyncLock:         s.syncLock,
+	}
+
+	// create new BABE session
+	bs, err := babe.NewSession(bsConfig)
+	if err != nil {
+		log.Error("[core] failed to initialize BABE session", "error", err)
+		return nil, err
+	}
+
+	log.Debug(
+		"[core] BABE session initialized",
+		"epoch", s.epochNumber,
+	)
+
+	return bs, nil
+}
+
+// handleBlockDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleBlockDigest(header *types.Header) (err error) {
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest is not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// cast decoded consensus digest item to conensus digest type
+			consensusDigest, ok := item.(*types.ConsensusDigest)
+			if !ok {
+				return errors.New("failed to cast DigestItem to ConsensusDigest")
+			}
+
+			err = s.handleConsensusDigest(header, consensusDigest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// handleConsensusDigest checks if the first block has been set for the current
+// epoch, if first block has not been set, assumes first block and sets epoch
+// data for the next epoch using NextEpochDescriptor in the ConsensusDigest
+func (s *Service) handleConsensusDigest(header *types.Header, digest *types.ConsensusDigest) error {
+
+	// check if first block has been set for current epoch
+	if s.firstBlock != nil {
```

Comment:
also check if `header` has a lower number than the current first block and is also in this epoch?

Reply:
Not sure what the check for a lower number would do if the fist block is reset for each epoch during finalize and initialize babe session.

Are you saying if their are two blocks claiming to be the first block and their epoch number matches the current epoch, we should favor the first block with the lower number...?

Also, the check for current epoch is 5 lines down from this check for first block. Should this be moved? Can you clarify?

Reply:
yes, if you receive multiple blocks that are both in the current epoch and contain the consensus digest, you should favour the one with the lower number. the one with the higher number is definitely not the first block in that epoch if there's another block also in that epoch with a lower number